bigquery_etl/schema/__init__.py

```diff
@@ -0,0 +1,152 @@
+"""Query schema."""
+
+import attr
+import json
+import yaml
+
+from pathlib import Path
+from tempfile import NamedTemporaryFile
+from typing import Dict, Any
+
+from bigquery_etl.dryrun import DryRun
+
+SCHEMA_FILE = "schema.yaml"
+
+
+@attr.s(auto_attribs=True)
+class Schema:
+    """Query schema representation and helpers."""
+
+    schema: Dict[str, Any]
+
+    @classmethod
+    def from_query_file(cls, query_file: Path):
+        if not query_file.is_file() or query_file.suffix != ".sql":
+            raise Exception(f"{query_file} is not a valid SQL file.")
+
+        schema = DryRun(str(query_file)).get_schema()
+        return cls(schema)
+
+    @classmethod
+    def from_schema_file(cls, schema_file: Path):
+        """Create schema from a yaml schema file."""
+        if not schema_file.is_file() or schema_file.suffix != ".yaml":
+            raise Exception(f"{schema_file} is not a valid YAML schema file.")
+
+        with open(schema_file) as file:
+            schema = yaml.load(file, Loader=yaml.FullLoader)
+            return cls(schema)
+
+    @classmethod
+    def from_json(cls, json_schema):
+        """Create schema from JSON object."""
+        return cls(json_schema)
+
+    @classmethod
+    def for_table(cls, project, dataset, table, partitioned_by=None):
+        """Get the schema for a BigQuery table."""
+        query = f"SELECT * FROM {project}.{dataset}.{table}"
+
+        if partitioned_by:
+            query += f" WHERE {partitioned_by} = DATE('2020-01-01')"
+
+        # write query to temporary file so it can get dry run
+        tmp = NamedTemporaryFile()
+        with open(tmp.name, "w") as f:
+            f.write(query)
+
+        try:
+            return cls(DryRun(str(tmp.name)).get_schema())
+        except Exception as e:
+            print(f"Cannot get schema for {project}.{dataset}.{table}: {e}")
+            return cls({"fields": []})
+
+    def merge(self, other: "Schema"):
+        """Merge another schema into the schema."""
+        self._traverse(
+            "root", self.schema["fields"], other.schema["fields"], update=True
+        )
+
+    def equal(self, other: "Schema") -> bool:
+        """Compare to another schema"""
+        try:
+            self._traverse(
+                "root", self.schema["fields"], other.schema["fields"], update=False
+            )
+        except Exception as e:
+            print(e)
+            return False
+
+        return True
+
+    def _traverse(self, prefix, columns, other_columns, update=False):
```

Comment:
This traversal approach is great. I feel confident we'll be able to evolve this to meet our needs.